src/crypto/crypto.cpp

```diff
@@ -749,4 +752,28 @@ POP_WARNINGS
     sc_sub(&h, &h, &sum);
     return sc_isnonzero(&h) == 0;
   }
+
+  void crypto_ops::derive_view_tag(const key_derivation &derivation, size_t output_index, view_tag &view_tag) {
+    struct {
+      char salt[8]; // view tag domain-separator
+      char output_index[(sizeof(size_t) * 8 + 6) / 7];
+    } buf;
+    memcpy(buf.salt, "view_tag", 8); // leave off null terminator
+    char *end = buf.output_index;
+    tools::write_varint(end, output_index);
+    assert(end <= buf.output_index + sizeof buf.output_index);
+
+    char siphash_key[16];
+    memcpy(&siphash_key, &derivation, 16);
+
+    // view_tag_full = H[siphash_key](salt, output_index)
+    unsigned char view_tag_full[8];
+    siphash(&buf, end - reinterpret_cast<char *>(&buf), siphash_key, view_tag_full, 8); // siphash result will be 8 bytes
+
+    memwipe(siphash_key, 16);
+
+    // only need a slice of view_tag_full to realize optimal perf/space efficiency
+    static_assert(sizeof(crypto::view_tag) <= sizeof(view_tag_full));
```

Comment:
The `message` parameter is only optional since c++17 and the codebase required minimum is c++14.